docs/source/cookbook/parsing.rst

```diff
@@ -0,0 +1,394 @@
+
+Once we write a lambda calculus parser, we will also see how we can take advantage of a
+powerful built in expression parser in Idris 2 to write a small calculator that should be
+smart enough to parse the following expression:
+
+.. code-block:: text
+
+    1 + 2 - 3 * 4 / 5
+
+Lexer
+-----
+
+The main lexer module is under ``Text.Lexer``. This module contains ``toTokenMap`` which is a function
+that converts a ``List (Lexer, k) -> TokenMap (Token k)`` where ``k`` is a token kind. This function
+could be used for simple lexer to token mappings. The module also includes high level lexers for
+specifying quantity and common programming primitives like ``alphas``, ``intLit``,
+``lineComment`` and ``blockComment``.
+
+The ``Text.Lexer`` module also reexports ``Text.Lexer.Core``, ``Text.Quantity`` and ``Text.Token``.
+
+``Text.Lexer.Core`` provides the building blocks of the lexer, including a type called
+``Recognise`` which is the underlying data type for the lexer. The other important function that this
+module provides is a ``lex`` which takes in a lexer and returns the tokens.
+
+``Text.Quantity`` provides a data type ``Quantity`` which can be used with certain lexers to specify
+how many times something is expected to appear.
+
+``Text.Token`` provides a data type ``Token`` that represents a parsed token, it's kind and the text.
+This module also provides an important interface called ``TokenKind`` which tells the lexer how to map
+token kinds to Idris 2 types and how to convert each kind from a string to a value.
+
+Parser
+------
+
+The main parser module is under ``Text.Parser``. This module contains different grammar parsers, the main
+one being ``match`` which takes a ``TokenKind`` and returns the value as defined in the ``TokenKind``
+interface. There are other grammar parsers as well, but for our example, we will only be using ``match``.
+
+The ``Text.Parser`` module reexports ``Text.Parser.Core``, ``Text.Quantity`` and ``Text.Token``.
+
+``Text.Parser.Core`` provides the building blocks of the parser, including a type called ``Grammar``
+which is the underlying data type for the parser. The other important function that this module provides
+is ``parse`` which takes in a ``Grammar`` and returns the parsed expression.
+
+We covered ``Text.Quantity`` and ``Text.Token`` in the Lexer section so we're not going to
+repeat what they do here.
+
+Lambda Calculus Lexer & Parser
+------------------------------
+
+.. code-block:: idris
+    :caption: LambdaCalculus.idr
+    :linenos:
+
+    import Data.List
+    import Data.List1
+    import Text.Lexer
+    import Text.Parser
+
+    %default total
+
+    data Expr = App Expr Expr | Abs String Expr | Var String | Let String Expr Expr
+
+    Show Expr where
+      show (App e1 e2) = "(" ++ show e1 ++ " " ++ show e2 ++ ")"
+      show (Abs v e) = "(\\" ++ v ++ "." ++ show e ++ ")"
+      show (Var v) = v
+      show (Let v e1 e2) = "let " ++ v ++ " = " ++ show e1 ++ " in " ++ show e2
```

Comment:
This should hopefully produce a nicer ouptut.
It's debatable whether we should bother given that the cookbook is about something else but the
output we get in the example program is a bit annoyingly overparenthesised so...
```suggestion
      showPrec d (App e1 e2) = showParens (d == App) (showPrec (User 0) e1 ++ " " ++ showPrec App e2)
      showPrec d (Abs v e) = showParens (d > Open) ("\\" ++ v ++ "." ++ show e)
      showPrec d (Var v) = v
      showPrec d (Let v e1 e2) = showParens (d > Open) ("let " ++ v ++ " = " ++ show e1 ++ " in " ++ show e2)
```